--- ../../pihpsdr-dl1ycf/src/exit_menu.c	2024-10-11 11:17:27.689097878 +0200
+++ exit_menu.c	2024-10-11 21:48:06.623562394 +0200
@@ -87,6 +87,7 @@
   _exit(0);
 }
 
+#ifndef __APPLE__
 // cppcheck-suppress constParameterCallback
 static gboolean reboot_cb (GtkWidget *widget, GdkEventButton *event, gpointer data) {
   stop_program();
@@ -100,6 +101,7 @@
   (void) system("shutdown -h -P now");
   _exit(0);
 }
+#endif
 
 void exit_menu(GtkWidget *parent) {
   dialog = gtk_dialog_new();
@@ -127,6 +129,7 @@
   GtkWidget *exit_b = gtk_button_new_with_label("Exit");
   g_signal_connect (exit_b, "button-press-event", G_CALLBACK(exit_cb), NULL);
   gtk_grid_attach(GTK_GRID(grid), exit_b, col, row, 1, 1);
+  #ifndef __APPLE__
   col++;
   GtkWidget *reboot_b = gtk_button_new_with_label("Reboot");
   g_signal_connect (reboot_b, "button-press-event", G_CALLBACK(reboot_cb), NULL);
@@ -135,6 +138,7 @@
   GtkWidget *shutdown_b = gtk_button_new_with_label("Shutdown");
   g_signal_connect (shutdown_b, "button-press-event", G_CALLBACK(shutdown_cb), NULL);
   gtk_grid_attach(GTK_GRID(grid), shutdown_b, col, row, 1, 1);
+  #endif
   gtk_container_add(GTK_CONTAINER(content), grid);
   sub_menu = dialog;
   gtk_widget_show_all(dialog);
