--- ../../pihpsdr-dl1ycf/src/transmitter.c	2024-10-11 11:17:27.716142303 +0200
+++ transmitter.c	2024-10-13 20:10:12.363213388 +0200
@@ -59,6 +59,7 @@
 
 #define min(x,y) (x<y?x:y)
 #define max(x,y) (x<y?y:x)
+#define USE_CFC
 
 //
 // CW pulses are timed by the heart-beat of the mic samples.
@@ -1957,13 +1958,34 @@
   // 0 dB, therefore, the auto-leveler is also automatically
   // activated when en/dis-abling the compressor(de-)
   //
-  SetTXACompressorRun(tx->id, tx->compressor);
-  SetTXACompressorGain(tx->id, tx->compressor_level);
-  SetTXAosctrlRun(tx->id, tx->compressor);
-  SetTXALevelerSt(tx->id, tx->compressor);
-  SetTXALevelerAttack(tx->id, 1);
-  SetTXALevelerDecay(tx->id, 500);
-  SetTXALevelerTop(tx->id, 6.0);
+  
+  // baseband compressor means speech processor in WDSP
+  SetTXACompressorGain(tx->id, tx->compressor_level); // define in TX menu
+  SetTXACompressorRun(tx->id, tx->compressor); // ON=1 or OFF=0, tick Compression ON/OFF in TX menu
+  t_print("%s: Compressor state %d with level %.1fdb\n", __FUNCTION__, tx->compressor, tx->compressor_level);
+  
+  // CESSB handling
+  if (tx->compressor == 1) { 
+   SetTXAosctrlRun(tx->id, tx->compressor);
+   t_print("%s: CESSB state ON\n", __FUNCTION__);
+  } else {
+   SetTXAosctrlRun(tx->id, 0);
+   t_print("%s: CESSB state OFF, because CESSB can be used if Compressor is ON\n", __FUNCTION__);
+  }
+  
+  // leveler
+  double Leveler_MaxGain = 6.0; // in db
+  int Leveler_Attack = 1; // in ms
+  int Leveler_Decay = 500; // in ms
+
+  SetTXALevelerAttack(tx->id, Leveler_Attack);
+  SetTXALevelerDecay(tx->id, Leveler_Decay);
+  #if defined (USE_CFC)
+    Leveler_MaxGain = 15.0;
+  #endif
+  SetTXALevelerTop(tx->id, Leveler_MaxGain); // set Leveler MaxGain
+  SetTXALevelerSt(tx->id, tx->compressor);   // fire up Leveler in conjunktion with the baseband compressor
+  t_print("%s: Leveler state %d with MaxGain %.1fdb, Attack %dms, Decay %dms\n", __FUNCTION__, tx->compressor, Leveler_MaxGain, Leveler_Attack, Leveler_Decay);
 }
 
 void tx_set_ctcss(const TRANSMITTER *tx) {
@@ -2007,9 +2029,66 @@
 }
 
 void tx_set_equalizer(TRANSMITTER *tx) {
-  int nfreq = tx->eq_tenband ? 10 : 4;
-  SetTXAEQProfile(tx->id, nfreq, tx->eq_freq, tx->eq_gain);
-  SetTXAEQRun(tx->id, tx->eq_enable);
+  /*
+     Patch by DL1BZ: Using the CFC from WDSP lib in piHPSDR
+  */
+ 
+  // CFC can only be used if USE_CFC defined as constant at begin og this file
+  // the TX-EQ function is not affected, but the CFC runs only if TX-EQ is active
+  #ifdef USE_CFC
+    // define 10 EQ freq for the CFC multiband compressor
+    double CFC_EQ_tenbands[10] = { 50, 150, 300, 500, 750, 1250, 1750, 2300, 2800, 3100 };
+    // define gain in db per EQ band as pre-compressor
+    // double CFC_preEQ_Level[10] = { 0, 0, 0, 3, 3, 6, 9, 9, 9, 9 }; // similiar to my Thetis
+    double CFC_preEQ_Level[10] = { 0, 0, 3, 3, 3, 6, 6, 6, 9, 9 }; // some correction for used mic with Mac
+    // define gain in db per freq as post-compressor
+    double CFC_postEQ_Level[10] = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
+    // define pre-amp level in db for pre-compressor
+    double CFC_preGain = 3;
+    // define post-amp level in db for post-compressor
+    double CFC_postGain = -9;
+    // we need to set a pointer per array
+    double* pCFC_EQ_tenbands = CFC_EQ_tenbands;
+    double* pCFC_preEQ_Level = CFC_preEQ_Level;
+    double* pCFC_postEQ_Level = CFC_postEQ_Level;
+  #endif
+
+  // the [0] element of the array tx->eq_freq is NOT USED, only [1]..[10] if EQ10band or [1]..[4] if EQ4band
+  // so we have an array with 11 instead of 10 (or 5 instead of 4) elements of the array tx->eq_freq
+  // the element [0] of array tx->eq_gain represent the pre-amp gain value of TX-EQ, the other elements are the gain for the EQ bands (4 or 10)
+  int nfreq = tx->eq_tenband ? 10 : 4;                      //using EQ4 or EQ10
+  SetTXAEQProfile(tx->id, nfreq, tx->eq_freq, tx->eq_gain); // load the profile into TX-EQ
+  
+  // CFC can only be used if USE_CFC defined as constant at begin og this file
+  // the TX-EQ function is not affected, but the CFC runs only if TX-EQ is active
+  #ifdef USE_CFC
+    // load pre-defined CFC profile CFC_EQ_tenbands, CFC_preEQ_Level, CFC_postEQ_Level
+    SetTXACFCOMPprofile(tx->id, 10, pCFC_EQ_tenbands, pCFC_preEQ_Level, pCFC_postEQ_Level);
+    // set pre-defined pre-amp level CFC_preGain in db for CFC
+    SetTXACFCOMPPrecomp(tx->id, CFC_preGain);
+    // set pre-defined post-amp level CFC_postGain in db for CFC
+    SetTXACFCOMPPrePeq(tx->id, CFC_postGain);
+  #endif
+
+  SetTXAEQRun(tx->id, tx->eq_enable); // fire up the TX-EQ if enabled
+  // print debug output
+  t_print("%s: nfreq=%d, EQ bands: %lu\n", __FUNCTION__, nfreq, sizeof(tx->eq_freq) / sizeof(tx->eq_freq[0]));
+  for (int i = 1; i < 11; i++) {
+      t_print("%s: EQ[%d] setting %.1fHz with gain %.1f\n", __FUNCTION__, i, tx->eq_freq[i], tx->eq_gain[i]);
+  }
+  t_print("%s: TX-EQ state %d with Gain=%.1fdb\n", __FUNCTION__, tx->eq_enable, tx->eq_gain[0]);
+  #ifdef USE_CFC
+    // fire up the CFC pre-compressor only if TX-EQ enabled
+    SetTXACFCOMPRun(tx->id, tx->eq_enable);
+    // fire up the CFC post-compressor only if TX-EQ enabled
+    SetTXACFCOMPPeqRun(tx->id, tx->eq_enable);
+    // debug output, write to log (print double values with 1 decimal places %f -> %.1f)
+    t_print("%s: number of CFC-EQ bands: %lu\n", __FUNCTION__, sizeof(CFC_EQ_tenbands) / sizeof(CFC_EQ_tenbands[0]));
+    for (int i = 0; i < 10; i++) {
+      t_print("%s: CFC-EQ[%d] setting %.1fHz with pre-gain %.1fdb, post-gain %.1fdb\n", __FUNCTION__, i, CFC_EQ_tenbands[i], CFC_preEQ_Level[i], CFC_postEQ_Level[i]);
+    }
+    t_print("%s: CFC state %d with preGain=%.1fdb postGain=%.1fdb\n", __FUNCTION__, tx->eq_enable, CFC_preGain, CFC_postGain);
+  #endif
 }
 
 void tx_set_fft_size(const TRANSMITTER *tx) {
